File: backend/services/scoring/structural.js
/**
 * structural.js — Deterministic structural score (0-1)
 *
 * Pure math. No API calls. No AI.
 * Only external dependency: db (SQLite, read-only here).
 *
 * Weights:
 *   budget   × 0.30
 *   franchise× 0.25
 *   genre    × 0.20
 *   talent   × 0.15
 *   mpaa     × 0.10
 */
import db from '../../db/database.js';

/* ------------------------------------------------------------------ */
/* Constants                                                          */
/* ------------------------------------------------------------------ */

const GENRE_SCORES = {
  'Action':         0.80,
  'Adventure':      0.80,
  'Animation':      0.75,
  'Family':         0.75,
  'Science Fiction':0.70,
  'Fantasy':        0.70,
  'Horror':         0.65,
  'Comedy':         0.60,
  'Thriller':       0.55,
  'Crime':          0.55,
  'Mystery':        0.50,
  'Music':          0.50,
  'Romance':        0.45,
  'War':            0.45,
  'Drama':          0.40,
  'Biography':      0.40,
  'History':        0.35,
  'Western':        0.35,
  'Documentary':    0.20,
};

const MPAA_SCORES = {
  'G':     0.70,
  'PG':    0.75,
  'PG-13': 0.85,
  'R':     0.60,
  'NC-17': 0.20,
  'NR':    0.40,
};

const MAJOR_STUDIO_SIGNALS = [
  'disney', 'marvel', 'warner', 'universal', 'paramount',
  'sony', 'lionsgate', 'netflix', 'apple', 'amazon', 'a24',
  'dreamworks', 'pixar', 'new line', 'columbia',
  '20th century', 'searchlight',
];

/* ------------------------------------------------------------------ */
/* Helpers                                                            */
/* ------------------------------------------------------------------ */

function safeParse(value, fallback) {
  if (value == null) return fallback;
  if (typeof value === 'object') return value;
  try { return JSON.parse(value); } catch { return fallback; }
}

function round4(n) {
  return Math.round(n * 10000) / 10000;
}

/* ------------------------------------------------------------------ */
/* Budget sub-score                                                   */
/* ------------------------------------------------------------------ */

function scoreBudget(film) {
  const budget = parseInt(film.budget, 10) || 0;
  const inferred = !!film.budget_inferred;

  let score;
  let tier;

  if (budget >= 150_000_000) {
    score = 1.00; tier = 'tentpole';
  } else if (budget >= 80_000_000) {
    score = 0.80; tier = 'major';
  } else if (budget >= 25_000_000) {
    score = 0.55; tier = 'mid';
  } else if (budget >= 5_000_000) {
    score = 0.30; tier = 'indie';
  } else if (budget > 0) {
    score = 0.15; tier = 'micro';
  } else if (budget === 0 && inferred) {
    score = 0.65; tier = 'unknown';
  } else {
    score = 0.10; tier = 'unknown';
  }

  return { score, tier, value: budget };
}

/* ------------------------------------------------------------------ */
/* Franchise sub-score (DB lookup)                                    */
/* ------------------------------------------------------------------ */

function scoreFranchise(film) {
  const collection = safeParse(film.belongs_to_collection, null);

  if (!collection || !collection.id) {
    return { score: 0.30, label: 'original_ip' };
  }

  // Query for prior films in the same collection with known revenue.
  // Uses json_extract (SQLite 3.38+); falls back to a JS-side filter if unavailable.
  let maxRevenue = 0;
  let priorFilmsFound = 0;

  try {
    const row = db.prepare(`
      SELECT MAX(revenue) AS max_rev, COUNT(*) AS cnt
      FROM films
      WHERE json_extract(belongs_to_collection, '$.id') = ?
        AND status = 'released'
        AND revenue > 0
        AND tmdb_id != ?
    `).get(collection.id, film.tmdb_id || 0);

    maxRevenue    = row?.max_rev ?? 0;
    priorFilmsFound = row?.cnt    ?? 0;
  } catch {
    // json_extract not available — filter in JS.
    const rows = db.prepare(
      `SELECT revenue, belongs_to_collection FROM films
       WHERE status = 'released' AND revenue > 0 AND tmdb_id != ?`
    ).all(film.tmdb_id || 0);

    for (const r of rows) {
      const col = safeParse(r.belongs_to_collection, null);
      if (col?.id === collection.id) {
        priorFilmsFound++;
        if ((r.revenue || 0) > maxRevenue) maxRevenue = r.revenue;
      }
    }
  }

  if (priorFilmsFound === 0) {
    return { score: 0.65, label: 'franchise_unverified' };
  }
  if (maxRevenue >= 500_000_000) {
    return { score: 0.95, label: 'mega_franchise' };
  }
  if (maxRevenue >= 100_000_000) {
    return { score: 0.80, label: 'major_franchise' };
  }
  return { score: 0.65, label: 'franchise' };
}

/* ------------------------------------------------------------------ */
/* Genre sub-score                                                    */
/* ------------------------------------------------------------------ */

function scoreGenre(film) {
  const genres = safeParse(film.genres, []);
  if (!genres.length) {
    return { score: 0.45, primary_genre: 'unknown' };
  }
  let best = { genre: genres[0], score: GENRE_SCORES[genres[0]] ?? 0.45 };
  for (const g of genres) {
    const s = GENRE_SCORES[g] ?? 0.45;
    if (s > best.score) best = { genre: g, score: s };
  }
  return { score: best.score, primary_genre: best.genre };
}

/* ------------------------------------------------------------------ */
/* Talent sub-score (DB lookup)                                       */
/* ------------------------------------------------------------------ */

function scoreTalent(film) {
  const director = safeParse(film.director, null);
  const cast     = safeParse(film.cast_top5, []);

  const DEFAULT = 0.35;

  // Director
  let directorScore = DEFAULT;
  let directorFound = false;
  if (director?.tmdb_person_id) {
    const row = db.prepare(
      'SELECT avg_ow_when_leading FROM talent_scores WHERE tmdb_person_id = ? AND role = ?'
    ).get(director.tmdb_person_id, 'director');
    if (row) {
      directorScore = Math.min(row.avg_ow_when_leading / 200_000_000, 1.0);
      directorFound = true;
    }
  }

  // Top 2 cast
  const top2 = cast.slice(0, 2).filter((c) => c.tmdb_person_id);
  let castScore = DEFAULT;
  let castFound = 0;
  if (top2.length > 0) {
    const ids = top2.map((c) => c.tmdb_person_id);
    const placeholders = ids.map(() => '?').join(',');
    const rows = db.prepare(
      `SELECT avg_ow_when_leading FROM talent_scores
       WHERE tmdb_person_id IN (${placeholders}) AND role = 'actor'`
    ).all(...ids);
    castFound = rows.length;
    if (rows.length > 0) {
      const avg = rows.reduce((s, r) => s + r.avg_ow_when_leading, 0) / rows.length;
      castScore = Math.min(avg / 150_000_000, 1.0);
    }
  }

  const score = round4(directorScore * 0.40 + castScore * 0.60);
  return { score, director_score: directorScore, cast_score: castScore, director_found: directorFound, cast_found: castFound };
}

/* ------------------------------------------------------------------ */
/* MPAA sub-score                                                     */
/* ------------------------------------------------------------------ */

function scoreMpaa(film) {
  const rating = film.mpaa_rating || null;
  const score  = MPAA_SCORES[rating] ?? 0.50;
  return { score, rating: rating ?? 'unknown' };
}

/* ------------------------------------------------------------------ */
/* Production company bonus                                           */
/* ------------------------------------------------------------------ */

function checkMajorStudio(film) {
  const companies = safeParse(film.production_companies, []);
  return companies.some((c) =>
    MAJOR_STUDIO_SIGNALS.some((sig) => (c.name ?? '').toLowerCase().includes(sig))
  );
}

/* ================================================================== */
/* computeStructuralScore(film)                                       */
/* ================================================================== */

export function computeStructuralScore(film) {
  // Sub-scores
  const budget   = scoreBudget(film);
  const franchise = scoreFranchise(film);
  const genre    = scoreGenre(film);
  const talent   = scoreTalent(film);
  const mpaa     = scoreMpaa(film);

  const majorStudio = checkMajorStudio(film);

  // Production company bonus: if budget unknown but major studio, floor budget_score at 0.65.
  const rawBudget = parseInt(film.budget, 10) || 0;
  if (rawBudget === 0 && majorStudio) {
    budget.score = Math.max(budget.score, 0.65);
  }

  const score = round4(
    budget.score    * 0.30 +
    franchise.score * 0.25 +
    genre.score     * 0.20 +
    talent.score    * 0.15 +
    mpaa.score      * 0.10
  );

  return {
    score,
    components: {
      budget:   { score: budget.score,   tier: budget.tier,              value: budget.value },
      franchise:{ score: franchise.score, label: franchise.label },
      genre:    { score: genre.score,    primary_genre: genre.primary_genre },
      talent:   {
        score:          talent.score,
        director_score: round4(talent.director_score),
        cast_score:     round4(talent.cast_score),
        director_found: talent.director_found,
        cast_found:     talent.cast_found,
      },
      mpaa:     { score: mpaa.score,     rating: mpaa.rating },
    },
    major_studio: majorStudio,
  };
}

File: backend/services/scoring/structural.py
import json
from ...db.database import db

GENRE_SCORES = {
    'Action': 0.80, 'Adventure': 0.80, 'Animation': 0.75, 'Family': 0.75,
    'Science Fiction': 0.70, 'Fantasy': 0.70, 'Horror': 0.65, 'Comedy': 0.60,
    'Thriller': 0.55, 'Crime': 0.55, 'Mystery': 0.50, 'Music': 0.50,
    'Romance': 0.45, 'War': 0.45, 'Drama': 0.40, 'Biography': 0.40,
    'History': 0.35, 'Western': 0.35, 'Documentary': 0.20,
}

MPAA_SCORES = {'G': 0.70, 'PG': 0.75, 'PG-13': 0.85, 'R': 0.60, 'NC-17': 0.20, 'NR': 0.40}

MAJOR_STUDIO_SIGNALS = [
    'disney', 'marvel', 'warner', 'universal', 'paramount',
    'sony', 'lionsgate', 'netflix', 'apple', 'amazon', 'a24',
    'dreamworks', 'pixar', 'new line', 'columbia',
    '20th century', 'searchlight',
]


def _safe_parse(v, fallback):
    if v is None:
        return fallback
    if isinstance(v, (dict, list)):
        return v
    try:
        return json.loads(v)
    except Exception:
        return fallback


def _r4(n):
    return round(n * 10000) / 10000


def _score_budget(film):
    budget = int(film.get('budget') or 0)
    inferred = bool(film.get('budget_inferred'))
    if budget >= 150_000_000:
        return {'score': 1.00, 'tier': 'tentpole', 'value': budget}
    if budget >= 80_000_000:
        return {'score': 0.80, 'tier': 'major',    'value': budget}
    if budget >= 25_000_000:
        return {'score': 0.55, 'tier': 'mid',      'value': budget}
    if budget >= 5_000_000:
        return {'score': 0.30, 'tier': 'indie',    'value': budget}
    if budget > 0:
        return {'score': 0.15, 'tier': 'micro',    'value': budget}
    if budget == 0 and inferred:
        return {'score': 0.65, 'tier': 'unknown',  'value': budget}
    return {'score': 0.10, 'tier': 'unknown', 'value': budget}


def _score_franchise(film):
    collection = _safe_parse(film.get('belongs_to_collection'), None)
    if not collection or not collection.get('id'):
        return {'score': 0.30, 'label': 'original_ip'}

    try:
        row = db.execute(
            '''SELECT MAX(revenue) AS max_rev, COUNT(*) AS cnt
               FROM films
               WHERE json_extract(belongs_to_collection, '$.id') = ?
                 AND status = 'released' AND revenue > 0 AND tmdb_id != ?''',
            (collection['id'], film.get('tmdb_id') or 0)
        ).fetchone()
        max_rev = row['max_rev'] or 0 if row else 0
        cnt = row['cnt'] or 0 if row else 0
    except Exception:
        rows = db.execute(
            "SELECT revenue, belongs_to_collection FROM films WHERE status='released' AND revenue>0 AND tmdb_id!=?",
            (film.get('tmdb_id') or 0,)
        ).fetchall()
        max_rev, cnt = 0, 0
        for r in rows:
            col = _safe_parse(r['belongs_to_collection'], None)
            if col and col.get('id') == collection['id']:
                cnt += 1
                if (r['revenue'] or 0) > max_rev:
                    max_rev = r['revenue']

    if cnt == 0:
        return {'score': 0.65, 'label': 'franchise_unverified'}
    if max_rev >= 500_000_000:
        return {'score': 0.95, 'label': 'mega_franchise'}
    if max_rev >= 100_000_000:
        return {'score': 0.80, 'label': 'major_franchise'}
    return {'score': 0.65, 'label': 'franchise'}


def _score_genre(film):
    genres = _safe_parse(film.get('genres'), [])
    if not genres:
        return {'score': 0.45, 'primary_genre': 'unknown'}
    best_genre, best_score = genres[0], GENRE_SCORES.get(genres[0], 0.45)
    for g in genres:
        s = GENRE_SCORES.get(g, 0.45)
        if s > best_score:
            best_genre, best_score = g, s
    return {'score': best_score, 'primary_genre': best_genre}


def _score_talent(film):
    director = _safe_parse(film.get('director'), None)
    cast = _safe_parse(film.get('cast_top5'), [])
    DEFAULT = 0.35

    director_score, director_found = DEFAULT, False
    if director and director.get('tmdb_person_id'):
        row = db.execute(
            'SELECT avg_ow_when_leading FROM talent_scores WHERE tmdb_person_id=? AND role=?',
            (director['tmdb_person_id'], 'director')
        ).fetchone()
        if row:
            director_score = min(row['avg_ow_when_leading'] / 200_000_000, 1.0)
            director_found = True

    top2 = [c for c in cast[:2] if c.get('tmdb_person_id')]
    cast_score, cast_found = DEFAULT, 0
    if top2:
        ids = [c['tmdb_person_id'] for c in top2]
        placeholders = ','.join('?' * len(ids))
        rows = db.execute(
            f"SELECT avg_ow_when_leading FROM talent_scores WHERE tmdb_person_id IN ({placeholders}) AND role='actor'",
            ids
        ).fetchall()
        cast_found = len(rows)
        if rows:
            avg = sum(r['avg_ow_when_leading'] for r in rows) / len(rows)
            cast_score = min(avg / 150_000_000, 1.0)

    score = _r4(director_score * 0.40 + cast_score * 0.60)
    return {'score': score, 'director_score': _r4(director_score),
            'cast_score': _r4(cast_score), 'director_found': director_found, 'cast_found': cast_found}


def _score_mpaa(film):
    rating = film.get('mpaa_rating')
    return {'score': MPAA_SCORES.get(rating, 0.50), 'rating': rating or 'unknown'}


def _check_major_studio(film):
    companies = _safe_parse(film.get('production_companies'), [])
    return any(
        sig in (c.get('name') or '').lower()
        for c in companies for sig in MAJOR_STUDIO_SIGNALS
    )


def compute_structural_score(film):
    budget    = _score_budget(film)
    franchise = _score_franchise(film)
    genre     = _score_genre(film)
    talent    = _score_talent(film)
    mpaa      = _score_mpaa(film)
    major_studio = _check_major_studio(film)

    if (int(film.get('budget') or 0) == 0) and major_studio:
        budget['score'] = max(budget['score'], 0.65)

    score = _r4(
        budget['score']    * 0.30 +
        franchise['score'] * 0.25 +
        genre['score']     * 0.20 +
        talent['score']    * 0.15 +
        mpaa['score']      * 0.10
    )

    return {
        'score': score,
        'components': {
            'budget':    {'score': budget['score'],    'tier': budget['tier'],    'value': budget['value']},
            'franchise': {'score': franchise['score'], 'label': franchise['label']},
            'genre':     {'score': genre['score'],     'primary_genre': genre['primary_genre']},
            'talent':    {'score': talent['score'],    'director_score': talent['director_score'],
                          'cast_score': talent['cast_score'], 'director_found': talent['director_found'],
                          'cast_found': talent['cast_found']},
            'mpaa':      {'score': mpaa['score'],      'rating': mpaa['rating']},
        },
        'major_studio': major_studio,
    }
